### ISSUE TITLE
use cloudfront for static content hosting

### ISSUE THREAD

**Issue by thewillhuang** [apex/up#635]
- cheaper (maybe?)
- no use of lambda calls
- no cold starts, website responds faster!
- no 6mb size limit

**Comment by tj** [comment]
dup of #62. I'd like to put CF infront of Lambda though so you can just utilize http caching for assets (with dynamic apps as well). Currently AWS isn't capable of a great fully static site since you need to manually invalidate files etc

**Comment by tj** [comment]
Actuallllly I'll re-open since the S3 implementation would be different

**Comment by thewillhuang** [comment]
awesome! thank you 👍

**Comment by jazen** [comment]
Would love to see that feature--ran into the size limit too a few days ago.

**Comment by Prinzhorn** [comment]
I'd love to see the following that I've done semi-manually in the past which also solves the issue with invalidations and a bunch of others:

`up` will deploy static files (e.g. from `/public`) to S3 with CloudFront in front and it will also deploy the dynamic application to lambda. Now here's the twist: I've always deployed my apps with a `STATIC_FILES_VERSION` env variable. So when I made a deployment I incremented the number and pushed the static files into S3 prefixed with the version. E.g. `/14/css/style.css`. After the static files have been deployed I deployed the application and updated the STATIC_FILES_VERSION env variable. This turned out to be the best thing ever:

1. No need for cache busting using stupid tricks like `style.css?13234523445` or answering support tickets with "have you tried clearing your cache?"
2. Clean semantic URL (query strings aren't CDN friendly anyway)
3. You get rollback capabilities for your static assets for free 😱

Instead of incrementing a number up could use the same identifier it already uses to identify different lambda deployments. It could fully transparently handle static files and rollbacks that way. In my pug templates I always have a helper for static files that basically concat STATIC_FILES_ENDPOINT and STATIC_FILES_VERSION. up could provide similar info to the application using env variable.s

**Comment by thewillhuang** [comment]
an easier way would be to just hash the file contents and have that as a manifest

**Comment by Prinzhorn** [comment]
> an easier way would be to just hash the file contents and have that as a manifest

@thewillhuang could you elaborate? What would the manifest contain exactly? How does this compare to what I suggested?

If you're using up pro you can likely implement what I've described above using `predeploy` which combines `up env get`, `up env set` and `aws s3 cp --recursive public s3://your-bucket...`

**Comment by idiomatic** [comment]
instead of hosting all of a static site in S3 (`{"type":"static"}`), how about just directory (`{"type": "server", "static":{"dir":"public", "s3-properties":...} }`)?

**Comment by edlefebvre** [comment]
@tj any news on this feature?

**Comment by tj** [comment]
@edlefebvre to be honest I'm not making nearly enough to justify adding new features to Up unfortunately, it was worth a try but didn't really work out financially

**Comment by edlefebvre** [comment]
Oh OK, sorry to hear that. Does it mean you will stop maintaining this repo entirely?

**Comment by tj** [comment]
Nope! I'll still maintain it, maybe add the small feature here and there, but nothing huge.

**Comment by edlefebvre** [comment]
OK, thank you for your answer :)

**Comment by kaihendry** [comment]
I've been successfully orchestrating static sites with https://github.com/aws-samples/aws-cdk-examples/tree/master/typescript/static-site/

**Comment by edlefebvre** [comment]
Thank you @kaihendry, I will have a look.
